src/mjan.c

```diff
@@ -0,0 +1,345 @@
+
+  return moves;
+}
+
+// enacts a move on a board, returning the resulting position.
+POSITION DoMove(POSITION position, MOVE move) {
+  char board[BOARDSIZE];
+  generic_hash_unhash(position, board);
+
+  board[DECODEMOVEJDX(move)] = board[DECODEMOVEIDX(move)];
+  board[DECODEMOVEIDX(move)] = ECHAR;
+
+  return generic_hash_hash(board, NEXTPLAYER(generic_hash_turn(position)));
+} 
+
+// parses a position and returns if the position is a loss, or undecided.
+//
+// example results:
+// 
+//  bwbw
+//  ----  -> undecided
+//  ----  ->
+//  wbwb
+//
+//  --bw      --bw
+//  bw--  ->  *w--  -> loss
+//  -bw-  ->  -*w-  ->
+//  w-b-      w-*-
+//
+//  -w-b      -*-b
+//  bwb-  ->  b*b-  -> loss
+//  -w--  ->  -*--  ->
+//  wb--      wb--
+VALUE Primitive(POSITION position) {
+  char board[BOARDSIZE];
+  generic_hash_unhash(position, board);
+
+  // checks all diags and vertical loss conditions + horizontal loss conditions for first two rows
+  for (int idx = 0; idx < BOARDSIZE / 2; idx ++) {
+    if (board[idx] == ECHAR)
+      continue;
+
+    if (board[idx] == board[idx + 4] && board[idx + 4] == board[idx + 8]) // down
+      return lose;
+
+    if (idx % 4 > 1 && board[idx] == board[idx - 1] && board[idx - 1] == board[idx - 2]) // left
+      return lose;
+
+    if (idx % 4 > 1 && board[idx] == board[idx + 3] && board[idx + 3] == board[idx + 6]) // left diagonal
+      return lose;
+
+    if (idx % 4 < 2 && board[idx] == board[idx + 1] && board[idx + 1] == board[idx + 2]) // right
+      return lose;
+
+    if (idx % 4 < 2 && board[idx] == board[idx + 5] && board[idx + 5] == board[idx + 10]) // right diagonal
+      return lose;
+  }
+
+  // check bottom 2 rows manually
+  if (board[9] != ECHAR && (board[9] == board[10]) && (board[8] == board[9] || board[10] == board[11]))
+    return lose;
+
+  if (board[13] != ECHAR && (board[13] == board[14]) && (board[12] == board[13] || board[14] == board[15]))
+    return lose;
+
+  return undecided;
+}
+
+// for symmetries, effectively blank
+POSITION GetCanonicalPosition(POSITION position) {
+  return position;
+}
+
+/*********** END SOLVING FUNCTIONS ***********/
+/*********** BEGIN TEXTUI FUNCTIONS ***********/
+
+// prints a visually digestable position representation
+// with legend and prediction.
+void PrintPosition(POSITION position, STRING playerName, BOOLEAN usersTurn) {
+  char board[BOARDSIZE];
+  generic_hash_unhash(position, board);
+
+  printf("\n         (  1  2  3  4 )           : %c %c %c %c\n", // row one
+	  board[0],
+	  board[1],
+	  board[2],
+	  board[3]
+  );
+	
+  printf("         (  5  6  7  8 )           : %c %c %c %c\n", // row two
+	  board[4],
+	  board[5],
+	  board[6],
+	  board[7]
+  );
+	
+  printf("LEGEND:  (  9 10 11 12 )   TOTAL:  : %c %c %c %c\n", // row three
+	  board[8],
+	  board[9],
+	  board[10],
+	  board[11]
+  );
+
+  printf("         ( 13 14 15 16 )           : %c %c %c %c %s\n\n", // row four
+	  board[12],
+	  board[13],
+	  board[14],
+    board[15],
```

Comment:
should there be a tab here or is it intentionally no tab?

Reply:
wow good find actually, will fix in  next commit